operator/pkg/controllers/configconnector/vpa_test.go

```diff
@@ -0,0 +1,244 @@
+						OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+							Type: "object",
+							Properties: map[string]apiextensionsv1.JSONSchemaProps{
+								"spec": {
+									Type:                   "object",
+									XPreserveUnknownFields: boolPtr(true),
+								},
+								"status": {
+									Type:                   "object",
+									XPreserveUnknownFields: boolPtr(true),
+								},
+							},
+						},
+					},
+					Subresources: &apiextensionsv1.CustomResourceSubresources{
+						Status: &apiextensionsv1.CustomResourceSubresourceStatus{},
+					},
+				},
+			},
+		},
+	}
+	if err := c.Create(ctx, vpaCRD); err != nil {
+		// Ignore if already exists (parallel tests might race)
+		if !apierrors.IsAlreadyExists(err) {
+			t.Fatalf("error creating VPA CRD: %v", err)
+		}
+	}
+
+	// Wait for CRD to be established
+	// In a real env we should wait, but envtest might be fast enough or we can retry.
+	// Let's just proceed and see if it works, or add a small retry loop for the first VPA creation.
+
+	// Ensure namespaces exist
+	testcontroller.EnsureNamespaceExists(c, k8s.CNRMSystemNamespace)
+	testcontroller.EnsureNamespaceExists(c, k8s.OperatorSystemNamespace)
+
+	// Create ConfigConnector
+	cc := &corev1beta1.ConfigConnector{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-kcc-vpa",
+		},
+		Spec: corev1beta1.ConfigConnectorSpec{
+			Mode:                 "cluster",
+			GoogleServiceAccount: "foo@bar.iam.gserviceaccount.com",
+		},
+	}
+	if err := c.Create(ctx, cc); err != nil {
+		t.Fatalf("error creating ConfigConnector: %v", err)
+	}
+
+	// Create ControllerResource with VPA enabled
+	vpaEnabled := true
+	cr := &customizev1beta1.ControllerResource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cnrm-controller-manager",
+		},
+		Spec: customizev1beta1.ControllerResourceSpec{
+			VerticalPodAutoscalerEnabled: &vpaEnabled,
+		},
+	}
+	if err := c.Create(ctx, cr); err != nil {
+		t.Fatalf("error creating ControllerResource: %v", err)
+	}
+
+	// Create VPA object with recommendations
+	vpaGVK := schema.GroupVersionKind{
+		Group:   "autoscaling.k8s.io",
+		Version: "v1",
+		Kind:    "VerticalPodAutoscaler",
+	}
+	vpa := &unstructured.Unstructured{}
+	vpa.SetGroupVersionKind(vpaGVK)
+	// Trigger reconciliation to create VPA
+	r := newConfigConnectorReconciler(c)
+	m := testcontroller.ParseObjects(ctx, t, testcontroller.GetClusterModeWorkloadIdentityManifest())
+	if err := handleLifecycles(ctx, t, r, cc, m); err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Apply customizations explicitly to trigger VPA creation
+	if err := r.applyCustomizations()(ctx, cc, m); err != nil {
+		t.Fatalf("unexpected error applying customizations: %v", err)
+	}
+
+	// Verify VPA object is created
+	vpaKey := client.ObjectKey{
+		Namespace: k8s.CNRMSystemNamespace,
+		Name:      "cnrm-controller-manager",
+	}
+
+	// Retry getting VPA in case CRD is not yet established or reconciliation is still in progress
+	for i := 0; i < 10; i++ {
```

Comment:
We should already have helper functions for this pattern. (Eg https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollUntilContextTimeout)